gcc-12 adds many more build-time warnings

On IRC today @khwilliamson  mentioned that he was seeing a surprising number of build-time warnings when compiling with gcc version 12.  (He stated that he normally compiles with g++.)  Last month, in https://github.com/Perl/perl5/issues/19776#issuecomment-1258084732, I noticed that there were large differences in the number of build-time warnings between different *sub-versions* of `gcc-12`, at least on FreeBSD.

Since I have upgraded my laptop to Ubuntu 22.04 LTS, I have more recent versions available.  I decided to install `gcc-12` and try it out.  On this version of Ubuntu, the *default* C-compiler is `gcc-11`.  Below are the results of building an unthreaded `perl` with these two different versions of `gcc` (as well as an earlier threaded build on FreeBSD).

**gcc-11**

```
$ gcc --version | head -1
gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
```
*Linux*
```
$ report-build-warnings 8aca9cadef.linux.unthreaded.gcc-11.maketp.output.txt.gz 
File:  8aca9cadef.linux.unthreaded.gcc-11.maketp.output.txt.gz
[no output]
```

**gcc-12**
```
$ gcc-12 --version | head -1
gcc-12 (Ubuntu 12.1.0-2ubuntu1~22.04) 12.1.0
```
*Linux*
```
$ report-build-warnings 8aca9cadef.linux.unthreaded.gcc-12.maketp.output.txt.gz 
File:  8aca9cadef.linux.unthreaded.gcc-12.maketp.output.txt.gz

  Waddress                                  42
  Warray-bounds                              3
```
*FreeBSD*
```
$ report-build-warnings b93391659d.freebsd.threaded.gcc12.maketp.output.txt.gz
File:  b93391659d.freebsd.threaded.gcc12.maketp.output.txt.gz

  Waddress                                  42
  Warray-bounds                              3
```

Below is the breakdown of the new-on-gcc-12 build-time warnings on Linux:

**-Waddress**
```
File:  8aca9cadef.linux.unthreaded.gcc-12.maketp.output.txt.gz

[
  {
    char   => 9,
    group  => "Waddress",
    line   => 1640,
    source => "gv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 25,
    group  => "Waddress",
    line   => 1846,
    source => "gv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 26,
    group  => "Waddress",
    line   => 1860,
    source => "gv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 19,
    group  => "Waddress",
    line   => 3308,
    source => "gv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 9,
    group  => "Waddress",
    line   => 8112,
    source => "regcomp.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the pointer operand in \xE2\x80\x98old_re->sv_u.svu_pv + (sizetype)Perl_ReANY((const REGEXP *)old_re)->pre_prefix\xE2\x80\x99 must not be NULL",
  },
  {
    char   => 12,
    group  => "Waddress",
    line   => 1703,
    source => "dump.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 12,
    group  => "Waddress",
    line   => 1718,
    source => "dump.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 9,
    group  => "Waddress",
    line   => 5942,
    source => "util.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 23,
    group  => "Waddress",
    line   => 1861,
    source => "mg.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 15,
    group  => "Waddress",
    line   => 1873,
    source => "mg.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 28,
    group  => "Waddress",
    line   => 428,
    source => "mro_core.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 26,
    group  => "Waddress",
    line   => 790,
    source => "mro_core.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 39,
    group  => "Waddress",
    line   => 906,
    source => "mro_core.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 38,
    group  => "Waddress",
    line   => 1166,
    source => "mro_core.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 42,
    group  => "Waddress",
    line   => 1442,
    source => "hv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 18,
    group  => "Waddress",
    line   => 1451,
    source => "hv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 14,
    group  => "Waddress",
    line   => 1540,
    source => "hv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 44,
    group  => "Waddress",
    line   => 2193,
    source => "hv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 31,
    group  => "Waddress",
    line   => 2195,
    source => "hv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 41,
    group  => "Waddress",
    line   => 2246,
    source => "hv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 48,
    group  => "Waddress",
    line   => 3711,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 22,
    group  => "Waddress",
    line   => 3720,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 26,
    group  => "Waddress",
    line   => 3733,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 12,
    group  => "Waddress",
    line   => 3791,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 24,
    group  => "Waddress",
    line   => 3941,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 26,
    group  => "Waddress",
    line   => 3954,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 25,
    group  => "Waddress",
    line   => 4388,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 27,
    group  => "Waddress",
    line   => 6514,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 50,
    group  => "Waddress",
    line   => 6514,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 20,
    group  => "Waddress",
    line   => 6749,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 33,
    group  => "Waddress",
    line   => 10038,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 21,
    group  => "Waddress",
    line   => 10415,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 12,
    group  => "Waddress",
    line   => 10730,
    source => "sv.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 15,
    group  => "Waddress",
    line   => 946,
    source => "pp.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 33,
    group  => "Waddress",
    line   => 971,
    source => "pp.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 19,
    group  => "Waddress",
    line   => 434,
    source => "scope.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 20,
    group  => "Waddress",
    line   => 1201,
    source => "scope.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 41,
    group  => "Waddress",
    line   => 1316,
    source => "scope.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 14,
    group  => "Waddress",
    line   => 1351,
    source => "pp_sys.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 21,
    group  => "Waddress",
    line   => 70,
    source => "module3.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98bufptr\xE2\x80\x99 will never be NULL",
  },
  {
    char   => 9,
    group  => "Waddress",
    line   => 8112,
    source => "re_comp.c",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the pointer operand in \xE2\x80\x98*__builtin_memcpy(&<Uff30>, &({...})) + (sizetype)Perl_ReANY((const REGEXP *)old_re)->pre_prefix\xE2\x80\x99 must not be NULL",
  },
  {
    char   => 17,
    group  => "Waddress",
    line   => 4313,
    source => "APItest.xs",
    text   => "the comparison will always evaluate as \xE2\x80\x98true\xE2\x80\x99 for the address of \xE2\x80\x98hek_key\xE2\x80\x99 will never be NULL",
  },
]

```
**-Warray-bounds**
```
File:  8aca9cadef.linux.unthreaded.gcc-12.maketp.output.txt.gz

[
  {
    char   => 26,
    group  => "Warray-bounds",
    line   => 440,
    source => "hv.h",
    text   => "array subscript [0, 9223372036854775807] is outside array bounds of \xE2\x80\x98char[0]\xE2\x80\x99",
  },
  {
    char   => 26,
    group  => "Warray-bounds",
    line   => 440,
    source => "hv.h",
    text   => "array subscript [0, 9223372036854775807] is outside array bounds of \xE2\x80\x98char[0]\xE2\x80\x99",
  },
  {
    char   => 26,
    group  => "Warray-bounds",
    line   => 440,
    source => "hv.h",
    text   => "array subscript [0, 9223372036854775807] is outside array bounds of \xE2\x80\x98char[0]\xE2\x80\x99",
  },
]
```
So, `gcc-12` is emitting a lot more build-time warnings than `gcc-11` (and, for that matter, many more than `clang-14`).  We should address these warnings -- if only because newer versions are likely to add still more build-time warnings.

I've been observing these same error for a while now but never got time to get into it. The main error is this one (in different places):
```
gv.c:1640:9: warning: the comparison will always evaluate as ‘true’ for the address of ‘hek_key’ will never be NULL [-Waddress]
 1640 |     if (!HvNAME_get(stash)) {
      |         ^
In file included from perl.h:4495,
                 from gv.c:41:
hv.h:60:17: note: ‘hek_key’ declared here
   60 |      char       hek_key[1];        /* variable-length hash key */
      |                 ^~~~~~~
```
It looks wrong to me (`HvNAME_get` calls `HEK_KEY` conditionally). I'm pretty sure it has the same cause as the array-bounds error though:
```
hv.h:440:26: warning: array subscript [0, 9223372036854775807] is outside array bounds of ‘char[0]’ [-Warray-bounds]
  440 | #define HEK_FLAGS(hek)  (*((unsigned char *)(HEK_KEY(hek))+HEK_LEN(hek)+1))
      |                         ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'm seeing similar erros for GvNAME and HvENAME, that's really the same issue.


I'm also seeing this one:
```
regcomp.c: In function ‘Perl_re_op_compile’:
regcomp.c:8112:9: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘old_re->sv_u.svu_pv + (sizetype)Perl_ReANY((const REGEXP *)old_re)->pre_prefix’ must not be NULL [-Waddress]
 8112 |         && RX_PRECOMP(old_re)
      |         ^~
```
`RX_PRECOMP` indeed looks like it will always be true, so one that should probably change.

I think we need to turn off the array-bounds warning, or learn how to
annotate our "extendible structs" with a char[1] or char[0] at the end.

The PRECOMP() one is on my todo list to remove.

On Fri, 4 Nov 2022 at 02:01, Leon Timmermans ***@***.***>
wrote:

> I've been observing these same error for a while now but never got time to
> get into it. The main error is this one (in different places):
>
> gv.c:1640:9: warning: the comparison will always evaluate as ‘true’ for the address of ‘hek_key’ will never be NULL [-Waddress]
>
>  1640 |     if (!HvNAME_get(stash)) {
>
>       |         ^
>
> In file included from perl.h:4495,
>
>                  from gv.c:41:
>
> hv.h:60:17: note: ‘hek_key’ declared here
>
>    60 |      char       hek_key[1];        /* variable-length hash key */
>
>       |                 ^~~~~~~
>
>
> It looks wrong to me (HvNAME_get calls HEK_KEY conditionally). I'm pretty
> sure it has the same cause as the array-bounds error though:
>
> hv.h:440:26: warning: array subscript [0, 9223372036854775807] is outside array bounds of ‘char[0]’ [-Warray-bounds]
>
>   440 | #define HEK_FLAGS(hek)  (*((unsigned char *)(HEK_KEY(hek))+HEK_LEN(hek)+1))
>
>       |                         ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>
>
> I'm seeing similar erros for GvNAME and HvENAME, that's really the same
> issue.
>
> I'm also seeing this one:
>
> regcomp.c: In function ‘Perl_re_op_compile’:
>
> regcomp.c:8112:9: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘old_re->sv_u.svu_pv + (sizetype)Perl_ReANY((const REGEXP *)old_re)->pre_prefix’ must not be NULL [-Waddress]
>
>  8112 |         && RX_PRECOMP(old_re)
>
>       |         ^~
>
>
> RX_PRECOMP indeed looks like it will always be true, so one that should
> probably change.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Perl/perl5/issues/20410#issuecomment-1302836626>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAZ5R3ZDUJYRQFII4S2FM3WGRN5TANCNFSM6AAAAAARINK5FQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

> It looks wrong to me (`HvNAME_get` calls `HEK_KEY` conditionally).

Unless you're assuming the compiler tracks the insides of a macro, it's seeing the HEK_KEY() being evaluated in a boolean context.  I suspect we wouldn't get this warning if HvNAME_get() was an inline function.

We can avoid this warning by adding a HvHasNAME() macro that does the conditional check only.

This also picked up:
```
dump.c: In function ‘Perl_do_gv_dump’:
dump.c:1705:12: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘(char *)sv->sv_any->xiv_u.xivu_namehek + 8’ must not be NULL [-Waddress]
 1705 |     if (sv && GvNAME(sv)) {
      |            ^~
dump.c: In function ‘Perl_do_gvgv_dump’:
dump.c:1720:12: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘(char *)sv->sv_any->xiv_u.xivu_namehek + 8’ must not be NULL [-Waddress]
 1720 |     if (sv && GvNAME(sv)) {
      |            ^~
```
which the compiler is correct to complain about - in a  -DDEBUGGING build GvNAME() asserts GvNAME_HEK()

I'm not sure about the HEK_FLAGS() warnings:
```
mro_core.c: In function ‘Perl_mro_method_changed_in’:
hv.h:441:26: warning: array subscript [0, 9223372036854775807] is outside array bounds of ‘unsigned char[0]’ [-Warray-bounds]
  441 | #define HEK_FLAGS(hek)  (*((unsigned char *)(HEK_KEY(hek))+HEK_LEN(hek)+1))
      |                         ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
embed.h:242:70: note: in definition of macro ‘hv_common’
  242 | e hv_common(a,b,c,d,e,f,g,h)      Perl_hv_common(aTHX_ a,b,c,d,e,f,g,h)
      |                                                                ^

hv.h:455:34: note: in expansion of macro ‘HEK_FLAGS’
  455 | #define HEK_UTF8(hek)           (HEK_FLAGS(hek) & HVhek_UTF8)
      |                                  ^~~~~~~~~
hv.h:554:56: note: in expansion of macro ‘HEK_UTF8’
  554 |      hv_common((hv), NULL, HEK_KEY(hek), HEK_LEN(hek), HEK_UTF8(hek),   \
      |                                                        ^~~~~~~~
mro_core.c:1333:23: note: in expansion of macro ‘hv_fetchhek’
 1333 |     SV ** const svp = hv_fetchhek(PL_isarev, HvENAME_HEK(stash), 0);
      |                       ^~~~~~~~~~~
```
especially since we use `-fno-strict-aliasing` (this is from @demerphq's #20513 PR).

The compiler can make a lot of assumptions, and I think in this case I think it's assuming that the HEK is the entire object, since we don't have a FAM, or a `char hek_key[1];` at the end, and that hence the implicit `unsigned char[]` is zero length.

--- a/dump.c
+++ b/dump.c
@@ -1702,7 +1702,7 @@ Perl_do_gv_dump(pTHX_ I32 level, PerlIO *file, const char *name, GV *sv)
     PERL_ARGS_ASSERT_DO_GV_DUMP;
 
     Perl_dump_indent(aTHX_ level, file, "%s = 0x%" UVxf, name, PTR2UV(sv));
-    if (sv && GvNAME(sv)) {
+    if (sv) {
         SV * const tmpsv = newSVpvs("");
         PerlIO_printf(file, "\t\"%s\"\n",
                               generic_pv_escape( tmpsv, GvNAME(sv), GvNAMELEN(sv), GvNAMEUTF8(sv) ));
@@ -1717,7 +1717,7 @@ Perl_do_gvgv_dump(pTHX_ I32 level, PerlIO *file, const char *name, GV *sv)
     PERL_ARGS_ASSERT_DO_GVGV_DUMP;
 
     Perl_dump_indent(aTHX_ level, file, "%s = 0x%" UVxf, name, PTR2UV(sv));
-    if (sv && GvNAME(sv)) {
+    if (sv) {
        SV *tmp = newSVpvs_flags("", SVs_TEMP);
         const char *hvname;
         HV * const stash = GvSTASH(sv);

--- a/ext/XS-APItest/APItest.xs
+++ b/ext/XS-APItest/APItest.xs
@@ -4310,7 +4310,7 @@ CODE:
 SV *
 HvENAME(HV *hv)
 CODE:
-    RETVAL = hv && HvENAME(hv)
+    RETVAL = hv && HvHasENAME(hv)
               ? newSVpvn_flags(
                   HvENAME(hv),HvENAMELEN(hv),
                   (HvENAMEUTF8(hv) ? SVf_UTF8 : 0)

--- a/gv.c
+++ b/gv.c
@@ -1628,7 +1628,7 @@ S_gv_stashpvn_internal(pTHX_ const char *name, U32 namelen, I32 flags)
     stash = GvHV(tmpgv);
     if (!(flags & ~GV_NOADD_MASK) && !stash) return NULL;
     assert(stash);
-    if (!HvNAME_get(stash)) {
+    if (!HvHasNAME(stash)) {
         hv_name_set(stash, name, namelen, flags & SVf_UTF8 ? SVf_UTF8 : 0 );
 
         /* FIXME: This is a repeat of logic in gv_fetchpvn_flags */
@@ -1834,7 +1834,7 @@ S_parse_gv_stash_name(pTHX_ HV **stash, GV **gv, const char **name,
 
                 if (!(*stash = GvHV(*gv))) {
                     *stash = GvHV(*gv) = newHV();
-                    if (!HvNAME_get(*stash)) {
+                    if (!HvHasNAME(*stash)) {
                         if (GvSTASH(*gv) == PL_defstash && *len == 6
                             && strBEGINs(*name, "CORE"))
                             hv_name_sets(*stash, "CORE", 0);
@@ -1848,7 +1848,7 @@ S_parse_gv_stash_name(pTHX_ HV **stash, GV **gv, const char **name,
                         mro_package_moved(*stash, NULL, *gv, 1);
                     }
                 }
-                else if (!HvNAME_get(*stash))
+                else if (!HvHasNAME(*stash))
                     hv_name_set(*stash, nambeg, name_cursor - nambeg, is_utf8);
             }
 
@@ -3298,7 +3298,7 @@ Perl_gv_handler(pTHX_ HV *stash, I32 id)
     U32 newgen;
     struct mro_meta* stash_meta;
 
-    if (!stash || !HvNAME_get(stash))
+    if (!stash || !HvHasNAME(stash))
         return NULL;
 
     stash_meta = HvMROMETA(stash);

--- a/hv.c
+++ b/hv.c
@@ -2286,8 +2286,8 @@ Perl_hv_undef_flags(pTHX_ HV *hv, U32 flags)
             mro_isa_changed_in(hv);
         if (PL_stashcache) {
             DEBUG_o(Perl_deb(aTHX_ "hv_undef_flags clearing PL_stashcache for effective name '%"
-                             HEKf "'\n", HEKfARG(HvENAME_HEK(hv))));
-            (void)hv_deletehek(PL_stashcache, HvENAME_HEK(hv), G_DISCARD);
+                             HEKf "'\n", HEKfARG(HvENAME_HEK_NN(hv))));
+            (void)hv_deletehek(PL_stashcache, HvENAME_HEK_NN(hv), G_DISCARD);
         }
       }
 
@@ -2300,8 +2300,8 @@ Perl_hv_undef_flags(pTHX_ HV *hv, U32 flags)
       {
         if (name && PL_stashcache) {
             DEBUG_o(Perl_deb(aTHX_ "hv_undef_flags clearing PL_stashcache for name '%"
-                             HEKf "'\n", HEKfARG(HvNAME_HEK(hv))));
-            (void)hv_deletehek(PL_stashcache, HvNAME_HEK(hv), G_DISCARD);
+                             HEKf "'\n", HEKfARG(HvNAME_HEK_NN(hv))));
+            (void)hv_deletehek(PL_stashcache, HvNAME_HEK_NN(hv), G_DISCARD);
         }
         hv_name_set(hv, NULL, 0, flags);
       }

--- a/hv.c
+++ b/hv.c
@@ -1439,7 +1439,7 @@ S_hv_delete_common(pTHX_ HV *hv, SV *keysv, const char *key, STRLEN klen,
         /* If this is a stash and the key ends with ::, then someone is 
          * deleting a package.
          */
-        if (sv && SvTYPE(sv) == SVt_PVGV && HvENAME_get(hv)) {
+        if (sv && SvTYPE(sv) == SVt_PVGV && HvHasENAME(hv)) {
                 gv = (GV *)sv;
                 if ((
                      (klen > 1 && key[klen-2] == ':' && key[klen-1] == ':')
@@ -1448,7 +1448,7 @@ S_hv_delete_common(pTHX_ HV *hv, SV *keysv, const char *key, STRLEN klen,
                     )
                  && (klen != 6 || hv!=PL_defstash || memNE(key,"main::",6))
                  && (stash = GvHV((GV *)gv))
-                 && HvENAME_get(stash)) {
+                 && HvHasENAME(stash)) {
                         /* A previous version of this code checked that the
                          * GV was still in the symbol table by fetching the
                          * GV with its name. That is not necessary (and
@@ -1537,7 +1537,7 @@ S_hv_delete_common(pTHX_ HV *hv, SV *keysv, const char *key, STRLEN klen,
         if (sv) {
             /* deletion of method from stash */
             if (isGV(sv) && isGV_with_GP(sv) && GvCVu(sv)
-             && HvENAME_get(hv))
+             && HvHasENAME(hv))
                 mro_method_changed_in(hv);
 
             if (d_flags & G_DISCARD) {
@@ -2190,9 +2190,9 @@ Perl_hfree_next_entry(pTHX_ HV *hv, STRLEN *indexp)
     array[*indexp] = HeNEXT(entry);
     ((XPVHV*) SvANY(hv))->xhv_keys--;
 
-    if (   PL_phase != PERL_PHASE_DESTRUCT && HvENAME(hv)
+    if (   PL_phase != PERL_PHASE_DESTRUCT && HvHasENAME(hv)
         && HeVAL(entry) && isGV(HeVAL(entry))
-        && GvHV(HeVAL(entry)) && HvENAME(GvHV(HeVAL(entry)))
+        && GvHV(HeVAL(entry)) && HvHasENAME(GvHV(HeVAL(entry)))
     ) {
         STRLEN klen;
         const char * const key = HePV(entry,klen);
@@ -2243,7 +2243,7 @@ Perl_hv_undef_flags(pTHX_ HV *hv, U32 flags)
        if they will be freed anyway. */
     /* note that the code following prior to hv_free_entries is duplicated
      * in sv_clear(), and changes here should be done there too */
-    if (PL_phase != PERL_PHASE_DESTRUCT && HvNAME(hv)) {
+    if (PL_phase != PERL_PHASE_DESTRUCT && HvHasNAME(hv)) {
         if (PL_stashcache) {
             DEBUG_o(Perl_deb(aTHX_ "hv_undef_flags clearing PL_stashcache for '%"
                              HEKf "'\n", HEKfARG(HvNAME_HEK(hv))));
@@ -2281,7 +2281,7 @@ Perl_hv_undef_flags(pTHX_ HV *hv, U32 flags)
       struct mro_meta *meta;
       const char *name;
 
-      if (HvENAME_get(hv)) {
+      if (HvHasENAME(hv)) {
         if (PL_phase != PERL_PHASE_DESTRUCT)
             mro_isa_changed_in(hv);
         if (PL_stashcache) {

--- a/hv.h
+++ b/hv.h
@@ -328,9 +328,10 @@ whether it is valid to call C<HvAUX()>.
 /* This macro may go away without notice.  */
 #define HvNAME_HEK(hv) \
         (HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name ? HvNAME_HEK_NN(hv) : NULL)
+#define HvHasNAME(hv) \
+        (HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name && HvNAME_HEK_NN(hv))
 #define HvNAME_get(hv) \
-        ((HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name && HvNAME_HEK_NN(hv)) \
-                         ? HEK_KEY(HvNAME_HEK_NN(hv)) : NULL)
+        (HvHasNAME(hv) ? HEK_KEY(HvNAME_HEK_NN(hv)) : NULL)
 #define HvNAMELEN_get(hv) \
         ((HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name && HvNAME_HEK_NN(hv)) \
                                  ? HEK_LEN(HvNAME_HEK_NN(hv)) : 0)
@@ -344,17 +345,18 @@ whether it is valid to call C<HvAUX()>.
   HvAUX(hv)->xhv_name_count == -1 ? NULL                              : \
                                     HvAUX(hv)->xhv_name_u.xhvnameu_name \
  )
+#define HvHasENAME_HEK(hv) \
+        (HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name)
 #define HvENAME_HEK(hv) \
-        (HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name ? HvENAME_HEK_NN(hv) : NULL)
+        (HvHasENAME_HEK(hv) ? HvENAME_HEK_NN(hv) : NULL)
+#define HvHasENAME(hv) \
+        (HvHasENAME_HEK(hv) && HvAUX(hv)->xhv_name_count != -1)
 #define HvENAME_get(hv) \
-   ((HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name && HvAUX(hv)->xhv_name_count != -1) \
-                         ? HEK_KEY(HvENAME_HEK_NN(hv)) : NULL)
+        (HvHasENAME(hv) ? HEK_KEY(HvENAME_HEK_NN(hv)) : NULL)
 #define HvENAMELEN_get(hv) \
-   ((HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name && HvAUX(hv)->xhv_name_count != -1) \
-                                 ? HEK_LEN(HvENAME_HEK_NN(hv)) : 0)
+        (HvHasENAME(hv) ? HEK_LEN(HvENAME_HEK_NN(hv)) : 0)
 #define HvENAMEUTF8(hv) \
-   ((HvHasAUX(hv) && HvAUX(hv)->xhv_name_u.xhvnameu_name && HvAUX(hv)->xhv_name_count != -1) \
-                                 ? HEK_UTF8(HvENAME_HEK_NN(hv)) : 0)
+        (HvHasENAME(hv) ? HEK_UTF8(HvENAME_HEK_NN(hv)) : 0)
 
 /*
  * HvKEYS gets the number of keys that actually exist(), and is provided

--- a/mg.c
+++ b/mg.c
@@ -1858,7 +1858,7 @@ Perl_magic_clearisa(pTHX_ SV *sv, MAGIC *mg)
         I32 items = AvFILLp((AV *)mg->mg_obj) + 1;
         while (items--) {
             stash = GvSTASH((GV *)*svp++);
-            if (stash && HvENAME(stash)) mro_isa_changed_in(stash);
+            if (stash && HvHasENAME(stash)) mro_isa_changed_in(stash);
         }
 
         return 0;
@@ -1870,7 +1870,7 @@ Perl_magic_clearisa(pTHX_ SV *sv, MAGIC *mg)
 
     /* The stash may have been detached from the symbol table, so check its
        name before doing anything. */
-    if (stash && HvENAME_get(stash))
+    if (stash && HvHasENAME(stash))
         mro_isa_changed_in(stash);
 
     return 0;

--- a/mro_core.c
+++ b/mro_core.c
@@ -1327,16 +1327,17 @@ via, C<mro::method_changed_in(classname)>.
 void
 Perl_mro_method_changed_in(pTHX_ HV *stash)
 {
-    const char * const stashname = HvENAME_get(stash);
-    const STRLEN stashname_len = HvENAMELEN_get(stash);
-
-    SV ** const svp = hv_fetchhek(PL_isarev, HvENAME_HEK(stash), 0);
-    HV * const isarev = svp ? MUTABLE_HV(*svp) : NULL;
-
     PERL_ARGS_ASSERT_MRO_METHOD_CHANGED_IN;
+
+    const char * const stashname = HvENAME_get(stash);
 
     if(!stashname)
         Perl_croak(aTHX_ "Can't call mro_method_changed_in() on anonymous symbol table");
+
+    const STRLEN stashname_len = HvENAMELEN_get(stash);
+
+    SV ** const svp = hv_fetchhek(PL_isarev, HvENAME_HEK_NN(stash), 0);
+    HV * const isarev = svp ? MUTABLE_HV(*svp) : NULL;
 
     /* Inc the package generation, since a local method changed */
     HvMROMETA(stash)->pkg_gen++;

--- a/mro_core.c
+++ b/mro_core.c
@@ -425,8 +425,8 @@ Perl_mro_get_linear_isa(pTHX_ HV *stash)
 
     if (meta->mro_which != &dfs_alg) { /* skip for dfs, for speed */
         SV * const namesv =
-            (HvENAME(stash)||HvNAME(stash))
-              ? newSVhek(HvENAME_HEK(stash)
+            (HvHasENAME_HEK(stash) || HvHasNAME(stash))
+              ? newSVhek(HvHasENAME_HEK(stash)
                           ? HvENAME_HEK(stash)
                           : HvNAME_HEK(stash))
               : NULL;
@@ -787,7 +787,7 @@ Perl_mro_package_moved(pTHX_ HV * const stash, HV * const oldstash,
     if(!(flags & 1)) {
         SV **svp;
         if(
-         !GvSTASH(gv) || !HvENAME(GvSTASH(gv)) ||
+         !GvSTASH(gv) || !HvHasENAME(GvSTASH(gv)) ||
          !(svp = hv_fetchhek(GvSTASH(gv), GvNAME_HEK(gv), 0)) ||
          *svp != (SV *)gv
         ) return;
@@ -903,7 +903,7 @@ S_mro_gather_and_rename(pTHX_ HV * const stashes, HV * const seen_stashes,
     HE *entry;
     I32 riter = -1;
     I32 items = 0;
-    const bool stash_had_name = stash && HvENAME(stash);
+    const bool stash_had_name = stash && HvHasENAME(stash);
     bool fetched_isarev = FALSE;
     HV *seen = NULL;
     HV *isarev = NULL;
@@ -1163,7 +1163,7 @@ S_mro_gather_and_rename(pTHX_ HV * const stashes, HV * const seen_stashes,
                             stashentry && *stashentry && isGV(*stashentry)
                          && (substash = GvHV(*stashentry))
                         )
-                     || (oldsubstash && HvENAME_get(oldsubstash))
+                     || (oldsubstash && HvHasENAME(oldsubstash))
                     )
                     {
                         /* Add :: and the key (minus the trailing ::)

--- a/pp.c
+++ b/pp.c
@@ -943,7 +943,7 @@ PP(pp_undef)
             /* undef *Pkg::meth_name ... */
             bool method_changed
              =   GvCVu((const GV *)sv) && (stash = GvSTASH((const GV *)sv))
-              && HvENAME_get(stash);
+              && HvHasENAME(stash);
             /* undef *Foo:: */
             if((stash = GvHV((const GV *)sv))) {
                 if(HvENAME_get(stash))
@@ -968,7 +968,7 @@ PP(pp_undef)
             /* undef *Foo::ISA */
             if( strEQ(GvNAME((const GV *)sv), "ISA")
              && (stash = GvSTASH((const GV *)sv))
-             && (method_changed || HvENAME(stash)) )
+             && (method_changed || HvHasENAME(stash)) )
                 mro_isa_changed_in(stash);
             else if(method_changed)
                 mro_method_changed_in(

--- a/pp_sys.c
+++ b/pp_sys.c
@@ -1348,7 +1348,7 @@ PP(pp_select)
     if (!egv)
         egv = PL_defoutgv;
     hv = isGV_with_GP(egv) ? GvSTASH(egv) : NULL;
-    gvp = hv && HvENAME(hv)
+    gvp = hv && HvHasENAME(hv)
                 ? (GV**)hv_fetch(hv, GvNAME(egv), HEK_UTF8(GvNAME_HEK(egv)) ? -GvNAMELEN(egv) : GvNAMELEN(egv), FALSE)
                 : NULL;
     if (gvp && *gvp == egv) {

--- a/scope.c
+++ b/scope.c
@@ -451,7 +451,7 @@ Perl_save_gp(pTHX_ GV *gv, I32 empty)
         HV * const stash = GvSTASH(gv);
         bool isa_changed = 0;
 
-        if (stash && HvENAME(stash)) {
+        if (stash && HvHasENAME(stash)) {
             if (memEQs(GvNAME(gv), GvNAMELEN(gv), "ISA"))
                 isa_changed = TRUE;
             else if (GvCVu(gv))
@@ -1171,7 +1171,7 @@ Perl_leave_scope(pTHX_ I32 base)
             HV * hv;
             a0 = ap[0]; a1 = ap[1]; a2 = ap[2];
             hv = GvSTASH(a0.any_gv);
-            if (hv && HvENAME(hv) && (
+            if (hv && HvHasENAME(hv) && (
                     (a2.any_sv && SvTYPE(a2.any_sv) == SVt_PVCV)
                  || (*a1.any_svp && SvTYPE(*a1.any_svp) == SVt_PVCV)
                ))
@@ -1286,7 +1286,7 @@ Perl_leave_scope(pTHX_ I32 base)
             had_method = cBOOL(GvCVu(a0.any_gv));
             gp_free(a0.any_gv);
             GvGP_set(a0.any_gv, (GP*)a1.any_ptr);
-            if ((hv=GvSTASH(a0.any_gv)) && HvENAME_get(hv)) {
+            if ((hv=GvSTASH(a0.any_gv)) && HvHasENAME(hv)) {
                 if (memEQs(GvNAME(a0.any_gv), GvNAMELEN(a0.any_gv), "ISA"))
                     mro_isa_changed_in(hv);
                 else if (had_method || GvCVu(a0.any_gv))

--- a/sv.c
+++ b/sv.c
@@ -3728,7 +3728,7 @@ S_glob_assign_glob(pTHX_ SV *const dsv, SV *const ssv, const int dtype)
         /* If source has a real method, then a method is
            going to change */
         else if(
-         GvCV((const GV *)ssv) && GvSTASH(dsv) && HvENAME(GvSTASH(dsv))
+         GvCV((const GV *)ssv) && GvSTASH(dsv) && HvHasENAME(GvSTASH(dsv))
         ) {
             mro_changes = 1;
         }
@@ -3737,7 +3737,7 @@ S_glob_assign_glob(pTHX_ SV *const dsv, SV *const ssv, const int dtype)
     /* If dest already had a real method, that's a change as well */
     if(
         !mro_changes && GvGP(MUTABLE_GV(dsv)) && GvCVu((const GV *)dsv)
-     && GvSTASH(dsv) && HvENAME(GvSTASH(dsv))
+     && GvSTASH(dsv) && HvHasENAME(GvSTASH(dsv))
     ) {
         mro_changes = 1;
     }
@@ -3750,7 +3750,7 @@ S_glob_assign_glob(pTHX_ SV *const dsv, SV *const ssv, const int dtype)
         if(memEQs(name, len, "ISA")
          /* The stash may have been detached from the symbol table, so
             check its name. */
-         && GvSTASH(dsv) && HvENAME(GvSTASH(dsv))
+         && GvSTASH(dsv) && HvHasENAME(GvSTASH(dsv))
         )
             mro_changes = 2;
         else {
@@ -3808,7 +3808,7 @@ S_glob_assign_glob(pTHX_ SV *const dsv, SV *const ssv, const int dtype)
     }
     else if(mro_changes == 3) {
         HV * const stash = GvHV(dsv);
-        if(old_stash ? (HV *)HvENAME_get(old_stash) : stash)
+        if(old_stash ? HvHasENAME(old_stash) : cBOOL(stash))
             mro_package_moved(
                 stash, old_stash,
                 (GV *)dsv, 0
@@ -3958,7 +3958,7 @@ Perl_gv_setref(pTHX_ SV *const dsv, SV *const ssv)
                    (len > 1 && name[len-2] == ':' && name[len-1] == ':')
                 || (len == 1 && name[0] == ':')
                 )
-             && (!dref || HvENAME_get(dref))
+             && (!dref || HvHasENAME(dref))
             ) {
                 mro_package_moved(
                     (HV *)sref, (HV *)dref,
@@ -3971,7 +3971,7 @@ Perl_gv_setref(pTHX_ SV *const dsv, SV *const ssv)
          && memEQs(GvNAME((GV*)dsv), GvNAMELEN((GV*)dsv), "ISA")
          /* The stash may have been detached from the symbol table, so
             check its name before doing anything. */
-         && GvSTASH(dsv) && HvENAME(GvSTASH(dsv))
+         && GvSTASH(dsv) && HvHasENAME(GvSTASH(dsv))
         ) {
             MAGIC *mg;
             MAGIC * const omg = dref && SvSMAGICAL(dref)
@@ -4432,7 +4432,7 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
                 if (reset_isa) {
                     HV * const stash = GvHV(dsv);
                     if(
-                        old_stash ? (HV *)HvENAME_get(old_stash) : stash
+                        old_stash ? HvHasENAME(old_stash) : cBOOL(stash)
                     )
                         mro_package_moved(
                          stash, old_stash,
@@ -6558,7 +6558,7 @@ S_anonymise_cv_maybe(pTHX_ GV *gv, CV* cv)
     }
 
     /* if not, anonymise: */
-    gvname = (GvSTASH(gv) && HvNAME(GvSTASH(gv)) && HvENAME(GvSTASH(gv)))
+    gvname = (GvSTASH(gv) && HvHasNAME(GvSTASH(gv)) && HvHasENAME(GvSTASH(gv)))
                     ? newSVhek(HvENAME_HEK(GvSTASH(gv)))
                     : newSVpvn_flags( "__ANON__", 8, 0 );
     sv_catpvs(gvname, "::__ANON__");
@@ -6793,7 +6793,7 @@ Perl_sv_clear(pTHX_ SV *const orig_sv)
         case SVt_PVGV:
             if (isGV_with_GP(sv)) {
                 if(GvCVu((const GV *)sv) && (stash = GvSTASH(MUTABLE_GV(sv)))
-                   && HvENAME_get(stash))
+                   && HvHasENAME(stash))
                     mro_method_changed_in(stash);
                 gp_free(MUTABLE_GV(sv));
                 if (GvNAME_HEK(sv))
@@ -10082,7 +10082,7 @@ Perl_sv_resetpvn(pTHX_ const char *s, STRLEN len, HV * const stash)
                 if (GvAV(gv)) {
                     av_clear(GvAV(gv));
                 }
-                if (GvHV(gv) && !HvNAME_get(GvHV(gv))) {
+                if (GvHV(gv) && !HvHasNAME(GvHV(gv))) {
                     hv_clear(GvHV(gv));
                 }
             }
@@ -10458,9 +10458,10 @@ Perl_sv_ref(pTHX_ SV *dst, const SV *const sv, const int ob)
         dst = sv_newmortal();
 
     if (ob && SvOBJECT(sv)) {
-        HvNAME_get(SvSTASH(sv))
-                    ? sv_sethek(dst, HvNAME_HEK(SvSTASH(sv)))
-                    : sv_setpvs(dst, "__ANON__");
+        if (HvHasNAME(SvSTASH(sv)))
+            sv_sethek(dst, HvNAME_HEK(SvSTASH(sv)));
+        else
+            sv_setpvs(dst, "__ANON__");
     }
     else {
         const char * reftype = sv_reftype(sv, 0);
@@ -10774,7 +10775,7 @@ S_sv_unglob(pTHX_ SV *const sv, U32 flags)
     SvREFCNT_inc_simple_void_NN(sv_2mortal(sv));
     if (GvGP(sv)) {
         if(GvCVu((const GV *)sv) && (stash = GvSTASH(MUTABLE_GV(sv)))
-           && HvNAME_get(stash))
+           && HvHasNAME(stash))
             mro_method_changed_in(stash);
         gp_free(MUTABLE_GV(sv));
     }

--- a/util.c
+++ b/util.c
@@ -5933,7 +5933,7 @@ S_gv_has_usable_name(pTHX_ GV *gv)
 {
     GV **gvp;
     return GvSTASH(gv)
-        && HvENAME(GvSTASH(gv))
+        && HvHasENAME(GvSTASH(gv))
         && (gvp = (GV **)hv_fetchhek(
                         GvSTASH(gv), GvNAME_HEK(gv), 0
            ))

